disk/disk.go

```diff
@@ -18,130 +17,160 @@ type Disk struct {
 	Root string
 }
 
+type BlockReaderStream struct {
+	blockIndex int
+	blockOffset int
+	file io.ReadSeeker
+}
+
+func (brs *BlockReaderStream) NextBlock() (*Block, error) {
+	block := newBlock()
+	err := readBlock(brs.file, block, brs.blockIndex)
+	block.left = brs.blockOffset
+	brs.blockOffset = 0
+	brs.blockIndex += 1
+	return block, err
+}
+
 // ReadAt reads up to len(p) bytes starting at byte offset off
 // from the File into p.
 // It returns the number of bytes read and an error, if any.
 func (d *Disk) ReadAt(name string, p []byte, off int64) (int, error) {
+	dataOffset := int(off)
 	name = path.Join(d.Root, name)
-
 	// nil or zero length payload
 	if len(p) == 0 {
 		return 0, nil
 	}
 
-	// block size
-	bsize := blockSize(name)
-	// payload size
-	psize := bsize - crc32Len
-
 	f, err := os.OpenFile(name, os.O_RDONLY, 0600)
 	if err != nil {
 		return 0, err
 	}
 	defer f.Close()
 
-	stIdx := off / psize
-	stOff := off - stIdx*psize
-	// how many bytes have been read
+	index, offset := blockIndexAndOffset(dataOffset)
+	stream := &BlockReaderStream{index, offset, f}
 	read := 0
 	for {
-		buf := make([]byte, bsize-crc32Len)
-		n, err := readBlock(f, buf, stIdx, bsize)
-		// If the read starts with a non-aligend position,
-		// only copy partial block
-		copied := copy(p, buf[stOff:n])
-		stOff = 0
+		block, err := stream.NextBlock()
+		copied := copy(p, block.GetPayload())
 		// We just copied some data into p, shrink p
 		p = p[copied:]
-		// We want to exit the loop for 3 cases:
-		// 1. There is an error reading block
-		// 2. We can't copy into p anymore -- p is filled up
-		if err != nil || len(p) == 0 {
-			return read + copied, err
+		read += copied
+		if err != nil {
+			return read, err
 		}
-		// 3. We read a partial block -- reach the end of the file
-		if n < bsize-crc32Len {
-			return read + copied, io.EOF
+		if len(p) == 0 {
+			return read, nil
+		}
+		if block.right < payloadSize  {
+			return read, io.EOF
 		}
-		read += copied
-		stIdx++
 	}
 }
 
+type BlockWriterStream struct {
+	blockOffset int
+	data []byte
+	file io.ReadWriteSeeker
+}
+
+// NextBlock gets the next block from the input stream
+func (bws *BlockWriterStream) NextBlock() (*Block, error) {
+	block := newBlock()
+	block.right = 0
+	// if this is a full block, return it
+	if bws.blockOffset == 0 && len(bws.data) >= payloadSize {
+		block.buf = bws.data[:payloadSize]
+		block.right = payloadSize
+		bws.data = bws.data[payloadSize:]
+		return block, nil
+	}
+	// if this is a partial block, copy data into it
+	payloadLen := payloadSize - bws.blockOffset
+	block.left = bws.blockOffset
+	if len(bws.data) > payloadLen {
+		block.Copy(bws.blockOffset, bws.data[:payloadLen])
+		bws.data = bws.data[payloadLen:]
+	} else {
+		block.Copy(bws.blockOffset, bws.data)
+		bws.data = nil
+	}
+	bws.blockOffset = 0
+	return block, nil
+}
+
+func (d *Disk) getDataLength(f *os.File) int {
```

Comment:
getDataSize?

// getDataSize returns the size of data in file (excluding the crc header size)